dask/dataframe/io/parquet/arrow.py

```diff
@@ -0,0 +1,344 @@
+import json
+
+import pandas as pd
+import pyarrow as pa
+import pyarrow.parquet as pq
+from ....delayed import delayed
+from ....bytes.core import get_pyarrow_filesystem
+from ....utils import natural_sort_key, getargspec
+from ..utils import _get_pyarrow_dtypes, _meta_from_dtypes
+from ...utils import clear_known_categories
+
+from .utils import _parse_pandas_metadata, _normalize_index_columns, Engine
+
+
+class ArrowEngine(Engine):
+    @staticmethod
+    def read_partition(fs, piece, columns, index, partitions, categories):
+
+        with fs.open(piece.path, mode="rb") as f:
+            table = piece.read(
+                columns=columns,
+                partitions=partitions,
+                use_pandas_metadata=True,
+                file=f,
+                use_threads=False,
+            )
+        df = table.to_pandas(categories=categories, use_threads=False)
+        if df.index.name:
+            df = df.reset_index()
+
+        return df[list(columns)]
+
+    @staticmethod
+    def read_metadata(
+        fs, paths, categories=None, index=None, gather_statistics=None, **kwargs
+    ):
+
+        # In pyarrow, the physical storage field names may differ from
+        # the actual dataframe names. This is true for Index names when
+        # PyArrow >= 0.8.
+        # We would like to resolve these to the correct dataframe names
+        # as soon as possible.
+
+        dataset = pq.ParquetDataset(
+            paths, filesystem=get_pyarrow_filesystem(fs)
+        )
+
+        if dataset.partitions is not None:
+            partitions = [
+                n for n in dataset.partitions.partition_names if n is not None
+            ]
+        else:
+            partitions = []
+
+        schema = dataset.schema.to_arrow_schema()
+        columns = None
+
+        has_pandas_metadata = (
+            schema.metadata is not None and b"pandas" in schema.metadata
+        )
+
+        if has_pandas_metadata:
+            pandas_metadata = json.loads(
+                schema.metadata[b"pandas"].decode("utf8")
+            )
+            index_names, column_names, storage_name_mapping, column_index_names = _parse_pandas_metadata(
+                pandas_metadata
+            )
+        else:
+            index_names = []
+            column_names = schema.names
+            storage_name_mapping = {k: k for k in column_names}
+            column_index_names = [None]
+
+        if index is None and len(index_names) > 0:
+            if len(index_names) == 1:
+                index = index_names[0]
+            else:
+                index = index_names
+
+        column_names += [p for p in partitions if p not in column_names]
```

Comment:
Just to confirm, you are recommending that we simply concatinate `column_names` and `partitions` without checking for overlap, correct?  (Since, as you correctly note, there *should* be no overlap.)

Reply:
It would be reasonable to explicitly check, and error if we find an overlap